securetar/__init__.py

```diff
@@ -212,76 +478,68 @@ def open(self) -> tarfile.TarFile:
-                **({"fileobj": self._fileobj} if self._fileobj else {}),
+                fileobj=self._fileobj,
+                **self._extra_tar_args,
             )
             return self._tar
 
-        # Encrypted/Decrypted Tarfile
-        self._open_file()
-        if self._mode == MOD_READ:
-            self.securetar_header = SecureTarHeader.from_bytes(self._file)
-            cipher_mode = CipherMode.DECRYPT
-            derived_key_material = self._root_key_context.restore_key_material(
-                header=self.securetar_header
-            )
-        else:
-            derived_key_material = self._root_key_context.derive_key_material(
-                self._derived_key_id
-            )
-            cbc_rand = derived_key_material.nonce
-            self.securetar_header = SecureTarHeader(cbc_rand, 0)
-            self._file.write(self.securetar_header.to_bytes())
-            cipher_mode = CipherMode.ENCRYPT
-        self._setup_cipher(cipher_mode, derived_key_material)
+        # When encrypted, we have root key context
+        if TYPE_CHECKING:
+            assert self._root_key_context is not None
+
+        # Open underlying file or stream
+        self._file = self._open_file()
+
+        # Set up cipher layer
+        cipher_mode = (
+            CipherMode.DECRYPT if self._mode == MOD_READ else CipherMode.ENCRYPT
+        )
+        self._cipher = _SecureTarCipher(
+            self._file,
+            cipher_mode,
+            root_key_context=self._root_key_context,
+            derived_key_id=self._derived_key_id,
+        )
+        self._cipher.open(plaintext_size=0)  # Size unknown for streaming write
+        if TYPE_CHECKING:
+            assert self._cipher.securetar_header is not None
 
+        # Write header for encryption
+        if cipher_mode == CipherMode.ENCRYPT:
+            self._file.write(self._cipher.securetar_header.to_bytes())
+
+        # Open tar with cipher as fileobj
         self._tar = tarfile.open(
-            fileobj=self,
+            fileobj=self._cipher,
             mode=self._tar_mode,
             dereference=False,
             bufsize=self._bufsize,
         )
         return self._tar
 
-    def _open_file(self) -> None:
+    def _open_file(self) -> IO[bytes] | BinaryIO:
         if self._fileobj:
             # If we have a fileobj, we don't need to open a file
-            self._file = self._fileobj
+            return self._fileobj
         else:
-            if not self._name:
-                raise ValueError("No filename or fileobj provided")
+            # We check in constructor that name and fileobj are not both None
+            if TYPE_CHECKING:
+                assert self._name is not None
             read_mode = self._mode.startswith("r")
             if read_mode:
                 file_mode: int = os.O_RDONLY
             else:
                 file_mode = os.O_WRONLY | os.O_CREAT
 
             fd = os.open(self._name, file_mode, 0o666)
```

Comment:
I'd add a comment why we need a file descriptor before opening the file instead of just opening the file.